### PR TITLE
For Home - Week 2

### PR DESCRIPTION
**Pull request by pvarna**

### CODE REVIEW

File: Week02-GradientDescent/task01.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import numpy as np
+import matplotlib.pyplot as plt
+
+
+class Value:
+    data: float = 0.0
```
Comment (SimeonHristov99) [inline Week02-GradientDescent/task01.py:6]: What is the reason for having this?

I would advice having static variables. If this is meant to be a default value, better to be passed in the constructor:

def __init__(self, data: float = 0.) -> None

File: Week02-GradientDescent/task02.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+class Value:
+    data: float = 0.0
+
+    def __init__(self, data: float) -> None:
+        self.data = data
+
+    def __repr__(self) -> str:
+        return f"Value(data={self.data})"
+
+    def __add__(self, other: "Value") -> "Value":
```
Comment (SimeonHristov99) [inline Week02-GradientDescent/task02.py:14]: Yes, sadly this is a strange way to handle referring to class in itself. More and more it seems that the following approach is becoming more popular (from __future__ import annotations):

```python
from __future__ import annotations


class Value:

    def __init__(self, data: float) -> None:
        self.data = data

    def __repr__(self) -> str:
        return f'Value(data={self.data})'

    def __add__(self, rhs: Value) -> Value:
        return Value(self.data + rhs.data)


def main() -> None:
    x = Value(2.0)
    y = Value(-3.0)
    result = x + y
    print(result)


if __name__ == '__main__':
    main()
```

File: Week02-GradientDescent/task04.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
+import numpy as np
+import matplotlib.pyplot as plt
+from typing import Set
+
+
+class Value:
+    data: float = 0.0
+    _prev: Set["Value"] = set()
```
Comment (SimeonHristov99) [inline Week02-GradientDescent/task04.py:8]: Better to be in the constructor with a default value. They (these static variables) are used in non-static functions, so might as well have them be part of the instance.

File: Week02-GradientDescent/task05.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
+import numpy as np
+import matplotlib.pyplot as plt
+from typing import Set
+
+
+class Value:
+    data: float = 0.0
+    _prev: Set["Value"] = set()
+    _op: chr = ''
```
Comment (SimeonHristov99) [inline Week02-GradientDescent/task05.py:9]: same as the above.

File: Week02-GradientDescent/task11.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+    return dot
+
+
+def update_values(nodes: Set[Value], learning_rate: float = 0.01) -> None:
+    for node in nodes:
+        # ???
+        node.data += learning_rate * node.gradient  # doing gradient ascent in order to meet the values from the example
+        # if we were doing gradient descent --> node.data -= learning_rate * node.gradient
+        # maybe the confusion comes from the fact that the error is negative, which is not natural
```
Comment (SimeonHristov99) [inline Week02-GradientDescent/task11.py:147]: Yep - that is correct. Sorry about this - I've elaborated in GitHub that this should be expected. By the way, if I can formulate the task in a better way, let me know - I'll appreciate it!

File: Week02-GradientDescent/task12.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+import graphviz
+from typing import Set, Tuple, List, TypeAlias
+
+
+class Value:
+    data: float = 0.0
+    _prev: Set["Value"] = set()
+    _op: chr = ''
+    gradient: float = 0.0
+    label: str = ""
```
Comment (SimeonHristov99) [inline Week02-GradientDescent/task12.py:12]: same as above. I'm starting to suspect this is done to have default value, but we can just move those default values in the constructor

File: Week02-GradientDescent/task12.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+    def __init__(self, data: float, label: str) -> None:
```
Comment (SimeonHristov99) [inline Week02-GradientDescent/task12.py:14]: actually, now we have two labels :sweat_smile: !

File: Week02-GradientDescent/task01.py (hunk 1 above)
Comment (pvarna) [reply Week02-GradientDescent/task01.py:6]: I though that it is like in C++ - you have to define all your data variables outside of the methods. I didn't know that it would work with only defining them in the constructor

File: Week02-GradientDescent/task05.py (hunk 4 above)
Comment (pvarna) [reply Week02-GradientDescent/task05.py:9]: Same as above

File: Week02-GradientDescent/task12.py (hunk 6 above)
Comment (pvarna) [reply Week02-GradientDescent/task12.py:12]: Same as above

File: Week02-GradientDescent/task11.py (hunk 5 above)
Comment (pvarna) [reply Week02-GradientDescent/task11.py:147]: I answered in the email

File: Week02-GradientDescent/task04.py (hunk 3 above)
Comment (pvarna) [reply Week02-GradientDescent/task04.py:8]: Same as above
